cmd/collector/app/handler.go

```diff
@@ -56,31 +63,71 @@ func (aH *APIHandler) RegisterRoutes(router *mux.Router) {
 }
 
 func (aH *APIHandler) saveSpan(w http.ResponseWriter, r *http.Request) {
+	bodyBytes, err := ioutil.ReadAll(r.Body)
+	r.Body.Close()
+	if err != nil {
+		http.Error(w, fmt.Sprintf(unableToReadBodyErrFormat, err), http.StatusInternalServerError)
+		return
+	}
+
 	format := r.FormValue(formatParam)
 	switch strings.ToLower(format) {
 	case "jaeger.thrift":
-		bodyBytes, err := ioutil.ReadAll(r.Body)
-		r.Body.Close()
-		if err != nil {
-			http.Error(w, fmt.Sprintf("Unable to read from body due to error: %v", err), http.StatusBadRequest)
-			return
-		}
-
 		tdes := thrift.NewTDeserializer()
 		// (NB): We decided to use this struct instead of straight batches to be as consistent with tchannel intake as possible.
 		var req tJaeger.CollectorSubmitBatchesArgs
 		if err = tdes.Read(&req, bodyBytes); err != nil {
-			http.Error(w, fmt.Sprintf("Cannot deserialize body due to error: %v", err), http.StatusBadRequest)
+			http.Error(w, fmt.Sprintf(unableToReadBodyErrFormat, err), http.StatusBadRequest)
 			return
 		}
 		ctx, cancel := tchanThrift.NewContext(time.Minute)
 		defer cancel()
 		if _, err = aH.jaegerBatchesHandler.SubmitBatches(ctx, req.Batches); err != nil {
-			http.Error(w, fmt.Sprintf("Cannot submit Jaeger batch due to error: %v", err), http.StatusInternalServerError)
-		} else {
-			w.WriteHeader(http.StatusOK)
+			http.Error(w, fmt.Sprintf("Cannot submit Jaeger batch: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+	case "zipkin.thrift":
+		spans, err := deserializeZipkin(bodyBytes)
+		if err != nil {
+			http.Error(w, fmt.Sprintf(unableToReadBodyErrFormat, err), http.StatusBadRequest)
+			return
+		}
+
+		ctx, _ := tchanThrift.NewContext(time.Minute)
+		if _, err = aH.zipkinSpansHandler.SubmitZipkinBatch(ctx, spans); err != nil {
+			http.Error(w, fmt.Sprintf("Cannot submit Zipkin batch: %v", err), http.StatusInternalServerError)
+			return
 		}
+
 	default:
 		http.Error(w, fmt.Sprintf("Unsupported format type: %v", format), http.StatusBadRequest)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func deserializeZipkin(b []byte) ([]*zipkincore.Span, error) {
+	buffer := thrift.NewTMemoryBuffer()
+	buffer.Write(b)
+
+	transport := thrift.NewTBinaryProtocolTransport(buffer)
+	_, size, err := transport.ReadListBegin() // Ignore the returned element type
+	if err != nil {
+		return nil, err
+	}
+
+	// We don't depend on the size returned by ReadListBegin to preallocate the array because it
+	// sometimes returns a nil error on bad input and provides an unreasonably large int for size
+	var spans []*zipkincore.Span
+	for i := 0; i < size; i++ {
+		zs := &zipkincore.Span{}
+		if err = zs.Read(transport); err != nil {
+			return nil, err
+		}
+		spans = append(spans, zs)
 	}
+
+	return spans, nil
```

Comment:
missing call ReadListEnd(). It does not matter if it does anything or not, that is its own implementation detail, but the Transport API says you need to call it at the end (maybe its behavior will change in future versions).

But you _can_ ignore the error returned from ReadListEnd(), since we've already read all spans anyway.